src/drivers/cmakeFileApiDriver.ts

```diff
@@ -221,36 +221,47 @@ export class CMakeFileApiDriver extends CMakeDriver {
         return 0;
     }
 
-    async doConfigure(args_: string[], outputConsumer?: proc.OutputConsumer, showCommandOnly?: boolean): Promise<number> {
-        const api_path = this.getCMakeFileApiPath();
+    async doConfigure(args_: string[], outputConsumer?: proc.OutputConsumer, showCommandOnly?: boolean, configurePreset?: ConfigurePreset | null): Promise<number> {
+        const api_path = this.getCMakeFileApiPath(configurePreset?.binaryDir);
         await createQueryFileForApi(api_path);
 
         // Dup args so we can modify them
         const args = Array.from(args_);
-
         // -S and -B were introduced in CMake 3.13 and this driver assumes CMake >= 3.15
         args.push(`-S${util.lightNormalizePath(this.sourceDir)}`);
-        args.push(`-B${util.lightNormalizePath(this.binaryDir)}`);
-
-        const gen = this.generator;
         let has_gen = false;
         for (const arg of args) {
             if (arg.startsWith("-DCMAKE_GENERATOR:STRING=")) {
                 has_gen = true;
             }
         }
-        if (!has_gen && gen) {
-            args.push('-G');
-            args.push(gen.name);
-            if (gen.toolset) {
-                args.push('-T');
-                args.push(gen.toolset);
+        if (configurePreset) {
+            args.push(`-B${util.lightNormalizePath(configurePreset.binaryDir ? configurePreset.binaryDir : this.binaryDir)}`);
+            if (!has_gen && configurePreset.generator) {
+                args.push('-G');
+                args.push(configurePreset.generator);
+                if (configurePreset.toolset) {
+                    args.push('-T');
+                    args.push(typeof(configurePreset.toolset) === "string" ? configurePreset.toolset as string : configurePreset.toolset.value || "");
+                }
             }
-            if (gen.platform) {
-                args.push('-A');
-                args.push(gen.platform);
+        } else {
+            args.push(`-B${util.lightNormalizePath(this.binaryDir)}`);
+            const generator = this.generator;
+            if (!has_gen && generator) {
```

Comment:
rather than have two copies of this code, I think it would be better to construct a `const generator` based off the data in the `configurePreset` and then have both paths run through the same code.